### PR TITLE
Add getting started section to the add page in developer perspective

### PR DESCRIPTION
**Pull request by jerolimov**
**Fixes**: 
Epic: https://issues.redhat.com/browse/ODC-5013
Story: https://issues.redhat.com/browse/ODC-5560

**Analysis / Root cause**: 
As a user, I want to see the new getting started section with areas to "Set up your cluster", Quick Starts, Guided Tours and more Learning resources.

**Solution Description**: 
This PR adds a new section to "+Add" page in the developer perspective above all options to import different types of resources. (This grid of add actions are redesigned as part of issue #8867.)

This new "Getting started" section adds three content cards with different links:

1. "Create applications using samples" - The first column is based on the developer catalog and shows two samples. The code contains two hardcoded "featured" samples (Quarkus and Sprint Boot) which should be preferred. When this samples are not found, some other samples will be displayed.
2. "Build with guided documentation" - This column shows up to two quick starts. It also prefers two hardcoded quickstarts (Quarkus and Sprint Boot again). It but will automatically hide them when the user completed them. When all quick started are completed, this column disappears.
3. "Explore new developer features" shows some static links. Currently it shows a link to `https://developers.redhat.com/products/openshift/getting-started` and two options to open the topology. The first link opens also opens the "Add to project" search bar.

It contains an option to hide this getting started section.

*Note:* The button too restore this section is not correct aligned, as this need some changes from #8867. Depending on which PR is merged first, this PR here or #8867 need to address this.

**Screen shots / Gifs for design review**: 
@openshift/team-devconsole-ux

image

Peek 2021-05-18 02-26

**Unit test coverage report**: 
image

image

**Test setup:**
* Open the developer perspective, Add Page.
* Check the different options.
* Complete a shown quick start and check that another quick start is shown.

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by jerolimov** [comment]
/kind feature

Since the epic acceptance criteria of [ODC-5013](https://issues.redhat.com/browse/ODC-5013) have not changed and the epic has `doc-ack` and `px-ack`, we approve this on behalf of them.
/label px-approved
/label docs-approved

Add hold to wait for an `qe-approved` also if someone adds a lgtm!
/hold

**Comment by sanketpathak** [comment]
Tested the PR through cluster-bot. Everything works as expected
Screenshot from 2021-05-18 14-06-44

Screenshot from 2021-05-18 14-19-50

/label qe-approved

**Comment by jerolimov** [comment]
Remove hold because this PR has now also approvals and just needs a lgtm.
/hold cancel

**Comment by jerolimov** [comment]
Fixed an i18n file issue and updated the "Explore new developer features" content based on latest feedback from Gina and Serena.

**Comment by gdoyle1** [comment]
@jerolimov Can you possibly help me get this up on my computer? I am trying to figure out if the font in the simple list is correct or not...

Also we spoke about this on slack about updating the right-most column for the Developer card to contain:
a) Discover certified Helm Charts --> link to the Helm Chart Catalog
b) Start building your application quickly in topology --> Topology with the search overlay to search
c) See what’s new in OpenShift 4.x --> (dynamic version) opens https://developers.redhat.com/products/openshift/getting-started

Will that be a change in this PR or a separate one?

**Comment by jerolimov** [comment]
@gdoyle1 The latest content changes are part of this PR. I just didn't update the screenshots above. I send you a message on slack how to test this locally. I will share a cluster with this PR later with your (as soon as the cluster is available).

File: frontend/packages/console-shared/src/components/getting-started/__tests__/QuickStartGettingStartedCard.data.tsx
Comment (sahil143) [inline frontend/packages/console-shared/src/components/getting-started/__tests__/QuickStartGettingStartedCard.data.tsx:5]: we already have mock data for quick starts in @console/app

Also, this file name can be `.ts` instead of `.tsx`

File: frontend/packages/console-shared/src/components/getting-started/__tests__/QuickStartGettingStartedCard.spec.tsx
Comment (sahil143) [inline frontend/packages/console-shared/src/components/getting-started/__tests__/QuickStartGettingStartedCard.spec.tsx:14]: Are you using `useLayoutEffect` somewhere and why is it assigned `useEffect`?

File: frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx
Comment (sahil143) [inline frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx:7]:
```suggestion
import { GettingStartedCard, ALL_NAMESPACES_KEY, useActiveNamespace, useOpenshiftVersion, GettingStartedLink } from '@console/shared';
```

File: frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx
Comment (sahil143) [inline frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx:13]: How about 
```suggestion
  const version = (useOpenshiftVersion() || '4.x')
```

File: frontend/packages/dev-console/src/components/add/GettingStartedSection.tsx
Comment (sahil143) [inline frontend/packages/dev-console/src/components/add/GettingStartedSection.tsx:8]:
```suggestion
} from '@console/shared';
```

File: frontend/packages/dev-console/src/components/add/SampleGettingStartedCard.tsx
Comment (sahil143) [inline frontend/packages/dev-console/src/components/add/SampleGettingStartedCard.tsx:7]: merge imports

File: frontend/packages/dev-console/src/components/add/SampleGettingStartedCard.tsx
Comment (sahil143) [inline frontend/packages/dev-console/src/components/add/SampleGettingStartedCard.tsx:10]: should be a relative import

**Review by sahil143 — CHANGES_REQUESTED** [review]
button is misaligned

Screenshot 2021-05-19 at 2 15 32 PM

File: frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx
Comment (jerolimov) [reply frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx]: Doesn't work here because I wanted to test this in the unit tests. But I removed this inline component now and just render the SimpleListItem in the loop of the GettingStartedCard.

See https://github.com/jerolimov/console/blob/odc-5680/frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx#L57-L81

File: frontend/packages/console-shared/src/components/getting-started/__tests__/QuickStartGettingStartedCard.data.tsx
Comment (jerolimov) [reply frontend/packages/console-shared/src/components/getting-started/__tests__/QuickStartGettingStartedCard.data.tsx:5]: Changed to `.ts`.

I only see some yaml data, but no mock data which includes the expected featured quick starts. I'm fine with changing this, but I also like to keep this test in depend from others.

File: frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx
Comment (sahil143) [inline frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx:25]: I'm not sure how `ALL_NAMESPACES_KEY` would work here. we show the project list page when all namespace is selected on add page.
This getting started card will only show up when we have an active namespace. I dont think you need to create all-namepaces url.

File: frontend/packages/console-shared/src/components/getting-started/__tests__/QuickStartGettingStartedCard.spec.tsx
Comment (jerolimov) [reply frontend/packages/console-shared/src/components/getting-started/__tests__/QuickStartGettingStartedCard.spec.tsx:14]: You're right, it was not used here and so I removed this here.

I keep that in `GettingStartedGrid.spec.tsx` and `RestoreGettingStartedButton.spec.tsx` and added a comment there why this was required. It was required for some PatternFly components (Popover and Label) which use `useLayoutEffect` internally. It removes only a warning while running the tests.

File: frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx
Comment (jerolimov) [reply frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx:13]: I would like to stay with an empty string and show only "What's new in OpenShift" in that case.

File: frontend/packages/dev-console/src/components/add/SampleGettingStartedCard.tsx
Comment (jerolimov) [reply frontend/packages/dev-console/src/components/add/SampleGettingStartedCard.tsx:7]: merged the getting-started imports

File: frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx
Comment (jerolimov) [reply frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx:27]: There is a small edge case: When the user has no access to any namespace, opens the topology, she / he still see the add page. It might be broken also in some other cases.

I will create an issue for that and will ask PM if we need this "feature". If we are sure that we have always a namespace defined, I'm with you that this condition is not required here.

File: frontend/packages/dev-console/src/components/add/GettingStartedSection.tsx
Comment (jerolimov) [reply frontend/packages/dev-console/src/components/add/GettingStartedSection.tsx:10]: I would like to stay with this import as most imports are using this schema at the moment.

File: frontend/packages/dev-console/src/components/add/__tests__/SampleGettingStartedCard.data.tsx
Comment (invincibleJai) [inline frontend/packages/dev-console/src/components/add/__tests__/SampleGettingStartedCard.data.tsx:1]: shall we keep file as `.ts`

File: frontend/packages/console-shared/src/components/getting-started/QuickStartGettingStartedCard.tsx
Comment (jerolimov) [reply frontend/packages/console-shared/src/components/getting-started/QuickStartGettingStartedCard.tsx:64]: @invincibleJai The reason for two pushs is that we want show all `IN_PROGRESS` first, and then `NOT_STARTED` quick starts. Your idea mix them together.

**Comment by jerolimov** [comment]
> button is misaligned
> 
> Screenshot 2021-05-19 at 2 15 32 PM

This is fixed now after I rebased this and integrated the `RestoreGettingStartedButton` into `AddPageLayout`

**Comment by jerolimov** [comment]
@sahil143 @invincibleJai Updated the unit tests and added some `// Workarounds...` because of the import changes. When importing all components via the `index.ts` some more mocks (also for unused components) are required. Just FYI.

File: frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx
Comment (nemesis09) [inline frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx:25]: is there a scenario where we might have an `href` but external is `false`. if not, can we not use `href` itself for the checks instead of the external prop?

File: frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx
Comment (nemesis09) [inline frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx:97]: could we use moreLink && (...elements) instead of a ternary operator?

**Comment by jerolimov** [comment]
Fixed translation position via `yarn i18n`

**Review by nemesis09 — COMMENTED** [review]
did not test it locally in details. just going through the code a bit.

File: frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx
Comment (nemesis09) [inline frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx:13]: i think `useActiveNamespace` is also used in  `SamplesGettingStartedCard` and both are being rendered in the `GettingStartedSection` component.
could we fetch the `activeNamespace` once in the `GettingStartedSection` and pass it on to `SamplesGettingStartedCard` and `DeveloperFeaturesGettingStartedCard.tsx` ?

File: frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx
Comment (jerolimov) [reply frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx:97]: Sure, this is possible and only a personal preference. There is one edge case (which is not relevant here) where `condition ? <Component /> : null` is better then `condition && <Component />`. When your condition checks for a length you need to ensure you write `array.length > 0 && <Component />` and not just `array.length && <Component />`. This would render "0' when the array is empty. :smirk:

File: frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx
Comment (jerolimov) [reply frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx:13]: For now, I want to keep these cards as separate as possible. Maybe we load them with the extensions mechanism in the future as well.

**Comment by jerolimov** [comment]
Changed one PF `Text` to `Title` as @nemesis09 recommended [in this comment here](https://github.com/openshift/console/pull/8950#discussion_r6358322420)

**Comment by invincibleJai** [comment]
@jerolimov should note be in new line here

image

**Comment by gdoyle1** [comment]
@invincibleJai @jerolimov yes "note" would ideally be on its own line!!

**Comment by jerolimov** [comment]
@gdoyle1 Are you fine with this?

image

**Comment by jerolimov** [comment]
> and below some translations are missing for samples, quickstarts item or is this as expected?

This are dynamic values, I think there are no translations for the samples and quick starts at the moment.

**Comment by gdoyle1** [comment]
> @gdoyle1 Are you fine with this?
> 
> image

yes LGTM!

File: frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx
Comment (invincibleJai) [reply frontend/packages/dev-console/src/components/add/DeveloperFeaturesGettingStartedCard.tsx:15]: @jerolimov I think it's good to be consistent as here we have `4.x` https://github.com/openshift/console/blob/master/frontend/packages/dev-console/src/components/guided-tour/GuidedTourText.tsx#L16

either way is fine but shouldn't it be same?

**Comment by invincibleJai** [comment]
there is an edge case as we discussed for self provisioner user, which has no projects have observed below scenarios

- on view all samples ulr is wrong says 404
- No project exists message hides the details on/off switch and the show getting started section again button (fixed)
- No project exists info alert has a strange bottom padding/margin

image

**Comment by jerolimov** [comment]
> there is an edge case as we discussed for self provisioner user, which has no projects have observed below scenarios
> 
> * on view all samples URL is wrong says 404
> * No project exists message hides the details on/off switch and the show getting started section again button
> * No project exists info alert has a strange bottom padding/margin

These 3 issues should be fixed now. As well as the latest border issue.

Without any project it looks like this now:
Screenshot from 2021-05-20 17-37-10

With at least one project available:
Screenshot from 2021-05-20 17-37-38

I will raise an issue if we should not show the add page anymore in topology (as we don't do this with 4.8 anymore when there is no resource available for the visualization).

**Comment by nemesis09** [comment]
> there is an edge case as we discussed for self provisioner user, which has no projects have observed below scenarios
> 
>     * on view all samples ulr is wrong says 404
> 
>     * No project exists message hides the details on/off switch and the show getting started section again button (fixed)
> 
>     * No project exists info alert has a strange bottom padding/margin
> 
> 
> image

I think the Empty State for the scenario where user doesn't have access to any projects, doesn't look super nice with both the hints. We should raise a separate issue for an Empty state, maybe similar to the one we already use for Toplogy

**Comment by invincibleJai** [comment]
> > there is an edge case as we discussed for self provisioner user, which has no projects have observed below scenarios
> > ```
> > * on view all samples ulr is wrong says 404
> > 
> > * No project exists message hides the details on/off switch and the show getting started section again button (fixed)
> > 
> > * No project exists info alert has a strange bottom padding/margin
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > image
> 
> I think the Empty State for the scenario where user doesn't have access to any projects, doesn't look super nice with both the hints. We should raise a separate issue for an Empty state, maybe similar to the one we already use for Topology.

yeah, I was discussing this with @jerolimov and we need to Jira ticket to track and discuss this with the design team.

**Comment by invincibleJai** [comment]
/lgtm

verified the changes and works fine for diff scenarios

**Pull request merged**
